qmail-remote.c

```diff
@@ -221,14 +275,44 @@ void smtp()
  
   if (smtpcode() != 220) quit("ZConnected to "," but greeting failed");
  
+#ifdef SMTPUTF8
+  /*-
+   * this part of the code is general purpose
+   * it can be used for
+   * checking other extensions like
+   * AUTH, STARTTLS, SIZE, etc
+   */
+  substdio_puts(&smtpto,"EHLO ");
+  substdio_put(&smtpto,helohost.s,helohost.len);
+  substdio_puts(&smtpto,"\r\n");
+  substdio_flush(&smtpto);
+
+  if (smtpcode() != 250) { /*- do a HELO if EHLO fails */
+    substdio_puts(&smtpto,"HELO ");
+    substdio_put(&smtpto,helohost.s,helohost.len);
+    substdio_puts(&smtpto,"\r\n");
+    substdio_flush(&smtpto);
+
+    code = smtpcode();
+    if (code >= 500) quit("DConnected to "," but my name was rejected");
+    if (code != 250) quit("ZConnected to "," but my name was rejected");
+  } else /* EHLO succeeded. Let's check SMTPUTF8 capa */
+    smtputf8 = get_capability("SMTPUTF8"); /*- did the remote server advertize SMTPUTF8 */
+#else
   substdio_puts(&smtpto,"HELO ");
   substdio_put(&smtpto,helohost.s,helohost.len);
   substdio_puts(&smtpto,"\r\n");
   substdio_flush(&smtpto);
   if (smtpcode() != 250) quit("ZConnected to "," but my name was rejected");
+#endif
  
   substdio_puts(&smtpto,"MAIL FROM:<");
   substdio_put(&smtpto,sender.s,sender.len);
+#ifdef SMTPUTF8
+  if (enable_utf8 && (flagutf8 || utf8read()))
+    substdio_puts(&smtpto,"> SMTPUTF8\r\n");
+  else
+#endif
```

Comment:
It's cheap to check these variables even in the case where we built without EAI.

Reply:
if we #ifdef enable_utf8 to 0 then this comes for free as the compiler would optimize it away.